app.go

```diff
@@ -453,28 +464,92 @@ func ValidateApp(opts ...Option) error {
 // registered via Invoke options. See the documentation of the App struct for
 // details on the application's initialization, startup, and shutdown logic.
 func New(opts ...Option) *App {
-	logger := fxlog.DefaultLogger(os.Stderr)
-
 	app := &App{
-		log:          logger,
+		log:          new(logBuffer),
 		startTimeout: DefaultTimeout,
 		stopTimeout:  DefaultTimeout,
 	}
 
 	for _, opt := range opts {
 		opt.apply(app)
 	}
-	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.log)}
 
 	app.container = dig.New(
 		dig.DeferAcyclicVerification(),
 		dig.DryRun(app.validate),
 	)
+	var connectLogger func(logger fxlog.Logger) error
+	if app.fxlogConstructor != nil {
+		// Since user supplied a custom logger, use a buffered logger to hold
+		// all messages until user supplied logger is instantiated. Then we flush
+		// those messages after fully constructing the custom logger.
+		var buffer logBuffer
+		app.log = &buffer
+		connectLogger = buffer.Connect // func(fxlog.Logger) error
+	} else {
+		// If user hasn't supplied a custom implementation of fxlog.Logger
+		// then use default one.
+		app.log = fxlog.DefaultLogger(os.Stderr)
+	}
 
 	for _, p := range app.provides {
 		app.provide(p)
 	}
 
+	if app.err != nil {
+		return app
+	}
+
+	// If WithLogger and Printer are both provided, WithLogger takes precedence.
+	// If user supplied an fxlog constructor, we need to have dig's container
+	// initialize it and grab all of arguments from the graph and then return.
+	if app.fxlogConstructor != nil {
+		// Here we add the custom logger into the container.
+		if err := app.container.Provide(app.fxlogConstructor); err != nil {
+			app.err = multierr.Combine(app.err,
+				fmt.Errorf("could not construct customer logger via fx.WithLogger: %w", err))
+			if connectLogger != nil {
+				err = connectLogger(fxlog.DefaultLogger(os.Stderr))
+				app.err = multierr.Combine(app.err, err)
+			} else {
```

Comment:
This will never be true unless there's a massive bug up above, in which case we should just panic. I would drop this else block and remove the `if` on connectLogger. Part of the contract of this code here is that if a custom constructor was provided, connectLogger was set.

Reply:
I have/had this check due to a nil deference linter, which points out that its possible to have connectLogger be nil here. I thought that we wanted to stay on the good side of it, but happy to remove this if you dont think this adds value.

Reply:
If we're trying to appease the linter here, I'm okay with a manual `panic` here because that's an irrecoverable "there;s a massive bug in Fx" sort of scenario.

Reply:
sg, though adding test coverage will be impossible so i changed my mind: its easier to ignore the lint warning.